contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -441,37 +441,112 @@ Alternatively, deployment could be simplified by having all components (includin
+
+#### ControllerPublishVolume/ControllerUnpublishVolume Credentials
+
+The CSI ControllerPublishVolume/ControllerUnpublishVolume calls are responsible for attaching and detaching volumes.
+These calls are executed by the CSI external-attacher.
+Credentials for these calls will be specified in the Kubernetes `CSIPersistentVolumeSource` object.
 
-// ...
 ```go
 type CSIPersistentVolumeSource struct {
 
-    // Optional: MountSecretRef is a reference to the secret object containing
-    // sensitive information to pass to the CSI driver during NodePublish.
+    // Optional: ControllerPublishSecretRef is a reference to the secret object
+    // containing sensitive information to pass to the CSI driver during
+    // ControllerPublishVolume and ControllerUnpublishVolume calls.
     // This may be empty if no secret is required. If the secret object contains
     // more than one secret, all secrets are passed.
     // +optional
-    MountSecretRef *SecretReference `json:"mountSecretRef,omitempty" protobuf:"bytes,3,opt,name=mountSecretRef"`
+    ControllerPublishSecretRef *SecretReference `json:"controllerPublishSecretRef,omitempty" protobuf:"bytes,4,opt,name=controllerPublishSecretRef"`
+}
+```
+
+If specified, the CSI external-attacher will fetch the Kubernetes secret referenced by `ControllerPublishSecretRef` and pass it to:
+1. The CSI `ControllerPublishVolume` in the `controller_publish_credentials` field.
+2. The CSI `ControllerUnpublishVolume` in the `controller_unpublish_credentials` field.
+
+See "Secret to CSI Credential Encoding" section above for details on how secrets will be mapped to CSI credentials.
+
+It is assumed that since `PersistentVolume` objects are non-namespaced and `CSIPersistentVolumeSource` can only be referenced via a `PersistentVolume`, only trusted users (e.g. cluster administrators) should be able to create a `PersistentVolume` objects and, thus, specify which secret to fetch.
+
+Permissions for the corresponding external-attacher, which would fetch this secret, may be limited to the specified namespace to prevent a compromised attacher from gaining any other secrets.
 
-    // Optional: AttachSecretRef is a reference to the secret object containing
-    // sensitive information to pass to the CSI driver during ControllerPublish.
+#### NodePublishVolume/NodeUnpublishVolume Credentials
+
+The CSI NodePublishVolume/NodeUnpublishVolume calls are responsible for mounting (setup) and unmounting (teardown) volumes.
+These calls are executed by the Kubernetes node agent (kubelet).
+Credentials for these calls will be specified in the Kubernetes `PersistentVolumeClaimSpec` object.
+
+```go
+type PersistentVolumeClaimSpec struct {
+
+    // Optional: NodePublishSecretRef is a reference to the secret object
+    // containing sensitive information to pass to the CSI driver during
```

Comment:
there has been much discussion of parameters on PVCs in the past, and this is essentially adding that via a referenced secret. for these user-provided credentials, it's worth discussing:
* how the user is expected to discover the structure of the parameters the provisioner expects
* the impact on portability of adding artifacts to a PVC (especially when combined with a default storage class)

Reply:
this covers user-provided credentials. is it common for the provisioner to need to make available a confidential parameter to the kubelet? if so, I would expect that to be provided on the PV.

Reply:
And it may help naming these better:

```go
type CSIPersistentVolumeSource struct {
  // Secret unique to this volume, e.g. encryption key.
  VolumeSecretRef *SecretReference

  // Secret unique per consumer, e.g. user credentials
  // This secret must exist in the namespace of the pod referencing this volume.
  WorkloadSecretRef *LocalObjectReference
}
```

Reply:
> remove NodePublishSecretRef *LocalObjectReference from PVC and move it to CSIPersistentVolumeSource

moving the reference to the PV makes sense to me

> with the caveat that it is still a LocalObjectReference and the namespace is assumed to be the namespace of the pod referencing the volume. This will permit both secrets that are unique per volume, as well as secrets that are unique per workload

I would strongly recommend keeping all the secret references in the PV namespace-qualified. That enables two use cases:
* a workload-specific secret living in the same namespace as the PVC
* a volume-specific secret living in a provisioner-managed namespace, hidden from the PVC (for example, a per-volume encryption key generated at provision time, placed in the provisioner-managed namespace, and made available to the kubelet responsible for mounting the volume)

Reply:
hmm, maybe I misread your suggestion (and yes, names are confusing :)

are PVCs from multiple namespaces allowed to reference a CSI PV?

Reply:
> a volume-specific secret living in a provisioner-managed namespace, hidden from the PVC (for example, a per-volume encryption key generated at provision time, placed in the provisioner-managed namespace, and made available to the kubelet responsible for mounting the volume)

Volume specific secret being fully qualified with namespace (`SecretReference`), I completely agree.

> a workload-specific secret living in the same namespace as the PVC

Workload specific secret being fully qualified with namespace, I'm not so sure about. If workload specific secret is a fully qualified with namespace (`SecretReference`), then regardless of the consumer, the secret is expected to exist in one namespace which the consumer may or may not have access to. Instead if workload secret is a `LocalObjectReference`, the namespace can be implied as always being the namespace of the caller. This allows for multiple consumers of a single PVC, each specifying their workload secret in their own namespace.

Reply:
> we effectively have three secrets in CSI:
>
> 1. Create/Delete Volume credential per class of volume
>    * Grants user of StorageClass ability to provision and delete volumes.
>    * Secret name/namespace in StorageClass makes sense.
> 2. Attach/Detach Volume credential per volume
>    * Grants all consumers access to volume, e.g. encryption key for volume
>    * Secret name/namespace in PV makes sense.
> 3. Mount/Unmount Volume credential per workload
>    * Grants a specific user of volume ability to consume volume: e.g. a single NFS share (represented by a PVC/PV) that permits multiple users access via different credentials.
>    * Secret name specified in PVC or pod would be nice, but as you stated that makes PVC non-portable.

A few questions:
* kubelets should only have access to `#3`, correct?
* is it odd to couple *when* a secret is needed to the location/visibility of the secret? don't we have secrets used at mount time that are set up by the provisioner today (azure file comes to mind)?

Reply:
@saad-ali Could you please confirm if `#3` has nothing to do with CSI, as I have a hard time mapping it to any CSI provided functions.

Reply:
> @saad-ali Could you please confirm if #3 has nothing to do with CSI, as I have a hard time mapping it to any CSI provided functions and if it is the case, should we limit this proposal to address just #1 and #2 which are relevant to CSI. What do you think?

@liggitt and I chatted offline. What I suggested in 3 won't work because a PVC/PV pair can only exist in a single namespace, and reusing the same volume requires creating a new PVC/PV pair, therefore it does not make sense to try and differentiate per volume vs per workload secrets. However, it enabling different secrets for different CSI operations should be allowed to allow a plugin to control which components have access to a secret. A secret required only on attach should not be required to be passed in on mount, for example (that would require kubelet to also have access to it when only external-attacher needs access).

Based on this discussion I will revise the proposal as follows:

```
type CSIPersistentVolumeSource struct {
  // Secrets required by the CSI driver to complete the ControllerPublishVolume call
  // Fetched by external-attacher.
  ControllerPublishSecretRef *SecretReference

  // Secrets required by the CSI driver to complete the NodeStageVolume call
  // Fetched by kubelet.
  NodeStageSecretRef *SecretReference

  // Secrets required by the CSI driver to complete the NodePublishVolume call
  // Fetched by kubelet.
  NodePublishSecretRef *SecretReference
}
```

All of these will be `SecretReference` with both namespace and name specified.

Reply:
that looks good to me:
* maps naturally to the CSI phases
* documents clearly which kubernetes component is expected to work with each piece
* provides the right information for the node authorizer to allow kubelets proper access
* supports a wide variety of use cases